### ISSUE TITLE
Correlate async dependencies with requests

### ISSUE THREAD

**Issue by abaranch** [Microsoft/ApplicationInsights-dotnet-server#113]
Currently it is a known limitation of the system.

**Comment by sescandell** [comment]
Hi,

Could this issue be the reason why one of our decorator (surrounding calls made to a ServiceBus) are, sometimes, not properly correlated to the "initial" request?

We see the telemetry in the data, but cannot see the tied request.

Here is the decorator:

```csharp
public class AIDependencyQueueWriterDecorator : IQueueWriter
    {
        private readonly TelemetryClient _telemetryClient;
        private readonly IQueueWriter _wrapped;
        private readonly string _queueName;

        public AIDependencyQueueWriterDecorator(TelemetryClient telemetryClient, IQueueWriter wrapped, string queueName)
        {
            _telemetryClient = telemetryClient;
            _wrapped = wrapped;
            _queueName = queueName;
        }

        public async Task<bool> WriteAsync<T>(T raw)
        {
            var stopWatch = new Stopwatch();
            var startTime = DateTime.UtcNow;
            stopWatch.Start();
            try
            {
                var res = await _wrapped.WriteAsync(raw);

                stopWatch.Stop();
                _telemetryClient.TrackDependency(new DependencyTelemetry
                {
                    Name = _queueName,
                    Target = _queueName,
                    Type = _wrapped.GetType().Name,
                    Duration = stopWatch.Elapsed,
                    Success = true,
                    ResultCode = "200 OK",
                    Timestamp = startTime
                });

                return res;
            }
            catch (Exception e)
            {
                stopWatch.Stop();
                _telemetryClient.TrackDependency(new DependencyTelemetry
                {
                    Name = _queueName,
                    Target = _queueName,
                    Type = _wrapped.GetType().Name,
                    Duration = stopWatch.Elapsed,
                    Success = false,
                    Properties = { {"Exception", e.Message} },
                    Timestamp = startTime
                });

                throw;
            }
        }
    }
```

Thanks,

**Comment by sescandell** [comment]
Hi again,

I think I found the issue in our context.

The thing is the application is not yet Async all the way down (from the controller to the inner services). Workaround is so to make some of these calls through a `Task.Run(WriteAsync(payload)).Wait();`. As a result, the CorrelationId is lost (as the HttpContext.Context used to get the platform context returns null into the CorrelationTelemetryInitializer).

Moving to an async from the controller to the inner call on a given action solved the issue in a MVC application (the thing is, a WebApi app using the same service project is not suffering this issue: I think because of OWIN integration).

I so see two things to fix this in our context:
 
 * using OWIN integration and tracking all dependencies and request through code with middleware
 * moving all the MVC code to async

If you can just confirm me `Task.Run().Wait()` might be the issue, that would be awesome.

Thanks,

**Comment by Dmitry-Matveev** [comment]
This bug is supposed to e fixed in 2.4. Please try it out.
